File: Data Structures & Algorithms/reorder-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void printList(ListNode head) {
         ListNode it = head;

         while(it != null) {
            System.out.print(it.val + " ");
            it= it.next;
        }
        System.out.println("\n");    
    }

    public void reorderList(ListNode head) {
        ListNode slow = head;
        ListNode fast = head.next;
        printList(head);

        // Split on 2 groups - in half
        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

       //System.out.println("slow: " + slow.val);
        //System.out.println("fast: " + fast.val);
    
        ListNode second = slow.next;
        slow.next = null;
        ListNode prev = null;

        while(second != null) {
            ListNode tmp = second.next;
            second.next = prev;
            prev = second;
            second = tmp;
        }

        ListNode first = head;
        second = prev;

        while(second != null) {
            ListNode tmp1 = first.next;
            ListNode tmp2 = second.next;

            first.next = second;
            second.next = tmp1;
            first = tmp1;
            second = tmp2;
        }
    
    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-1.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var dummy: ListNode = ListNode(-1)
        var current: ListNode = dummy

        var pos1 = l1
        var pos2 = l2

        var nrPos = 0
        var carry=0

        while (pos1 != null || pos2 != null || carry != 0) {
            val sum =  (pos1?.`val` ?: 0) + (pos2?.`val` ?: 0) + carry

            carry = sum / 10

           // println("$nrPos -> pos1: ${pos1?.`val`}, pos2: ${pos2?.`val`}, sum: $sum carry: $carry, pos1.next: ${pos1?.next?.`val`}, pos2.next: ${pos2?.next?.`val`}")

            current.next = ListNode(sum % 10)
            current = current.next!!

            pos1 = pos1?.next
            pos2 = pos2?.next
        }

        return dummy.next
    }
}
